File: frontend/de/unika/ipd/grgen/ast/containers/ArrayVarNode.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 5.0
 * Copyright (C) 2003-2020 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

/**
 * @author Edgar Jakumeit
 */

package de.unika.ipd.grgen.ast.containers;

import de.unika.ipd.grgen.ast.*;
import de.unika.ipd.grgen.ast.exprevals.*;
import de.unika.ipd.grgen.ir.exprevals.Expression;
import de.unika.ipd.grgen.ir.IR;
import de.unika.ipd.grgen.ir.containers.ArrayVarExpr;
import de.unika.ipd.grgen.parser.Coords;

public class ArrayVarNode extends ArrayAccumulationMethodNode
{
	static {
		setName(ArrayVarNode.class, "array var");
	}

	public ArrayVarNode(Coords coords, ExprNode targetExpr)
	{
		super(coords, targetExpr);
	}

	@Override
	protected boolean checkLocal() {
		TypeNode targetType = targetExpr.getType();
		if(!(targetType instanceof ArrayTypeNode)) {
			targetExpr.reportError("This argument to array var method must be of type array<T>");
			return false;
		}
		ArrayTypeNode arrayType = (ArrayTypeNode)targetExpr.getType();
		if(!arrayType.valueType.isAccumulatableType()) {
			targetExpr.reportError("The array value type of the array var method must be one of: " + TypeNode.getAccumulatableTypesAsString());
			return false;
		}
		return true;
	}

	@Override
	public TypeNode getType() {
		return BasicTypeNode.doubleType;
	}

	@Override
	public boolean isValidTargetTypeOfAccumulation(TypeNode type) {
		return type.isEqual(BasicTypeNode.doubleType);
	}

	@Override
	protected IR constructIR() {
		return new ArrayVarExpr(targetExpr.checkIR(Expression.class));
	}
}

File: frontend/de/unika/ipd/grgen/ast/expr/array/ArrayVarNode.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

/// <summary>
/// @author Edgar Jakumeit
/// </summary>

namespace de.unika.ipd.grgen.ast.expr.array
{
using ExprNode = de.unika.ipd.grgen.ast.expr.ExprNode;
using TypeNode = de.unika.ipd.grgen.ast.type.TypeNode;
using BasicTypeNode = de.unika.ipd.grgen.ast.type.basic.BasicTypeNode;
using ArrayTypeNode = de.unika.ipd.grgen.ast.type.container.ArrayTypeNode;
using IR = de.unika.ipd.grgen.ir.IR;
using Expression = de.unika.ipd.grgen.ir.expr.Expression;
using ArrayVarExpr = de.unika.ipd.grgen.ir.expr.array.ArrayVarExpr;
using Coords = de.unika.ipd.grgen.parser.Coords;

public class ArrayVarNode : ArrayAccumulationMethodNode
{
	static ArrayVarNode()
	{
		SetClassName(typeof(ArrayVarNode), "array var");
	}

	public ArrayVarNode(Coords coords, ExprNode targetExpr)
		: base(coords, targetExpr)
	{
	}

	protected internal override bool CheckLocal()
	{
		// target type already checked during resolving into this node
		ArrayTypeNode arrayType = TargetTypeExact;
		if(!arrayType.valueType.IsAccumulatableType())
		{
			targetExpr.ReportError("The array function method var can only be employed on an object of type array<" + TypeNode.AccumulatableTypesAsString + ">"
					+ " (but is employed on an object of type " + arrayType.TypeName + ").");
			return false;
		}
		return true;
	}

	public override TypeNode Type
	{
		get
		{
		return BasicTypeNode.doubleType;
		}
	}

	public override bool IsValidTargetTypeOfAccumulation(TypeNode type)
	{
		return type.IsEqual(BasicTypeNode.doubleType);
	}

	public override string ValidTargetTypesOfAccumulation
	{
		get
		{
		return "double";
		}
	}

	protected internal override IR ConstructIR()
	{
		targetExpr = targetExpr.Evaluate();
		return new ArrayVarExpr(targetExpr.CheckIR(typeof(Expression)));
	}
}

}
